roles/rsyslog/roles/input_roles/basics/tasks/main.yaml

```diff
@@ -1,10 +1,9 @@
 ---
-- name: Set basics facts
-  set_fact:
-    rsyslog_role_packages: "{{rsyslog_basics_packages | flatten | d([])}}"
-    rsyslog_role_rules: "{{ rsyslog_basics_rules }}"
-
+# Deploy configuration files
 - name: Install/Update basics role packages and generate example configuration files to role subdir
+  vars:
+    __rsyslog_packages: "{{ __rsyslog_basics_packages }}"
```

Comment:
@nhosoi you don't need faltten?

Reply:
Basically, as long as one level list ['package0', 'package1', ...] is passed, flatten is not needed.  You might have noticed that the variables like `__rsyslog_basic_packages` start with `__` that means they are internal set in each vars/main.yaml files.  We do not allow users to set them.  So, it's guaranteed that the passed lists are one level.

But there is one exception. :)  We allow users to install/update extra rsyslog packages with rsyslog_extra_packages here.
https://github.com/linux-system-roles/logging/pull/87/files#diff-ed5a08a2e61430dc18e2d3820cbd6f32R3
As you suggested, we'd better flatten rsyslog_extra_packages.  For instance, if this kind of nested list is given, it fails in the task.  I'm adding the flatten filter to rsyslog_extra_packages.  Thanks!!
rsyslog_extra_packages: [['rsyslog-mysql'], ['rsyslog-kafka']]

Reply:
@nhosoi why should such a nested list appear there?

Reply:
> @nhosoi why should such a nested list appear there?

@pcahyna, if we allow users to set a variable, any input could come in, so we'd better be prepared, I thought.

Reply:
@nhosoi I don't think we should expect any input, we should expect correct input. (Will we flatten all the lists passed to the role? If not, why is the package list special? If yes, will we do it in all the roles? We should as the roles should be consistent.)

Reply:
@pcahyna, in my past experiences, I learned users could make any mistakes and ask supports for help; to reduce the calls, we are supposed to add any efforts as much as possible. :)  Let's assume a user put an invalid list.  I guess we need to choose one of these?
1. with flatten, logging role successfully works.
2. with some list inspection code (which does not exist now, I think), logging role fails with a descriptive error.
3. without any support, logging role fails with a cryptic error...
@richm, @sradco, can I have your thoughts?

Reply:
Having some sort of input normalization/validation for this parameter establishes a precedent, that we will have input normalization/validation for _all_ input parameters for _all_ system roles.  I don't think we want that.  I don't think we need that, except in very specific cases where:
* it is easy for users to make a mistake
* the mistake could be very costly (e.g. some sort of destructive operation that loses data)

I don't think this qualifies, so I would prefer not to have any extra code that makes this more complex.

Reply:
I don't think we need a decision now. Please do not add new instances of `flatten`, but keep the existing ones. This is a topic for standards discussion.

Reply:
@richm I agree. Moreover, input normalization for cases where the mistake could be very costly is even less justified, because it could convert user input to something that the user has not intended (OTOH input validation would be justified.)

Reply:
@pcahyna, @richm, thanks for your discussions.  Yes, it'd be nice to have some standards for handling possible user mistakes.